drivers/rtc/rtc_atmel_sam.c

```diff
@@ -0,0 +1,651 @@
+		timalr |= RTC_TIMALR_SECEN;
+		timalr |= RTC_TIMALR_SEC(timeptr->tm_sec);
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_MINUTE) {
+		timalr |= RTC_TIMALR_MINEN;
+		timalr |= RTC_TIMALR_MIN(timeptr->tm_min);
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_HOUR) {
+		timalr |= RTC_TIMALR_HOUREN;
+		timalr |= RTC_TIMALR_HOUR(timeptr->tm_hour);
+	}
+
+	return timalr;
+}
+
+static uint32_t rtc_atmel_calalr_from_tm(const struct tm *timeptr, uint32_t mask)
+{
+	uint32_t calalr = 0;
+
+	if (mask & RTC_ALARM_TIME_MASK_MONTH) {
+		calalr |= RTC_CALALR_MTHEN;
+		calalr |= RTC_CALALR_MONTH(timeptr->tm_mon + 1);
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_MONTHDAY) {
+		calalr |= RTC_CALALR_DATEEN;
+		calalr |= RTC_CALALR_DATE(timeptr->tm_mday);
+	}
+
+	return calalr;
+}
+
+static uint32_t rtc_atmel_sam_alarm_mask_from_timalr(uint32_t timalr)
+{
+	uint32_t mask = 0;
+
+	if (timalr & RTC_TIMALR_SECEN) {
+		mask |= RTC_ALARM_TIME_MASK_SECOND;
+	}
+
+	if (timalr & RTC_TIMALR_MINEN) {
+		mask |= RTC_ALARM_TIME_MASK_MINUTE;
+	}
+
+	if (timalr & RTC_TIMALR_HOUREN) {
+		mask |= RTC_ALARM_TIME_MASK_HOUR;
+	}
+
+	return mask;
+}
+
+static uint32_t rtc_atmel_sam_alarm_mask_from_calalr(uint32_t calalr)
+{
+	uint32_t mask = 0;
+
+	if (calalr & RTC_CALALR_MTHEN) {
+		mask |= RTC_ALARM_TIME_MASK_MONTH;
+	}
+
+	if (calalr & RTC_CALALR_DATEEN) {
+		mask |= RTC_ALARM_TIME_MASK_MONTHDAY;
+	}
+
+	return mask;
+}
+
+static void rtc_atmel_sam_tm_from_timalr_calalr(struct tm *timeptr, uint32_t mask,
+						 uint32_t timalr, uint32_t calalr)
+{
+	memset(timeptr, 0x00, sizeof(*timeptr));
+
+	if (mask & RTC_ALARM_TIME_MASK_SECOND) {
+		timeptr->tm_sec = RTC_REG_GET_FIELD_BCD(timalr, TIMALR_SEC);
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_MINUTE) {
+		timeptr->tm_min= RTC_REG_GET_FIELD_BCD(timalr, TIMALR_MIN);
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_HOUR) {
+		timeptr->tm_hour = RTC_REG_GET_FIELD_BCD(timalr, TIMALR_HOUR);
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_MONTHDAY) {
+		timeptr->tm_mday = RTC_REG_GET_FIELD_BCD(calalr, CALALR_DATE);
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_MONTH) {
+		timeptr->tm_mon = RTC_REG_GET_FIELD_BCD(calalr, CALALR_MONTH) - 1;
+	}
+}
+
+static void rtc_atmel_sam_isr(const struct device *dev)
+{
+	struct rtc_atmel_sam_data *data = (struct rtc_atmel_sam_data *)dev->data;
+	uint32_t sr;
+
+	sr = REG_RTC_SR;
+
+	/* Update acknowledge */
+	if (sr & RTC_SR_ACKUPD) {
+		k_sem_give(&data->cr_upd_ack_sem);
+
+		REG_RTC_SCCR = RTC_SCCR_ACKCLR;
+	}
+
+	/* Alarm */
+	if ((sr & RTC_SR_ALARM) && (data->callback != NULL)) {
+		data->callback(dev, 0, data->user_data);
+
+		REG_RTC_SCCR = RTC_SCCR_ALRCLR;
+	}
+}
+
+/*************************************************************************************************/
+/*                                    RTC API implementation                                     */
+/*************************************************************************************************/
+static int rtc_atmel_sam_time_set(const struct device *dev, const struct tm *timeptr)
```

Comment:
You need check function names because it should be `rtc_sam_<verb>_<seomthing>`. See
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/counter/counter_sam_tc.c as reference,